### PR TITLE
[WIP] Poetry setup for PyElastica (#101)

### LINKED ISSUE #101: Setup pyproject (poetry)
For convenient package management, following `MiV-OS`, we should consider moving from `setup.py` to `poetry` i.e. `pyproject.toml` file setup.

### PR DESCRIPTION
**Pull request by bhosale2**
Fixes #101, a brief checklist:

- [x] Add `pyproject.toml/lock` and modify `main.yml`
- [x] Update `publish-to-pypi.yml`
- [x] update docs CI
- [x] update contrib guidelines to reflect poetry changes
- [x] update docs readme to reflect poetry changes
- [x] Remove `setup.py`, `requirements.txt` after confirming complete transfer of information

### CODE REVIEW

File: .github/workflows/ci-sphinx.yml [hunk 1]
```diff
@@ -36,10 +38,4 @@ jobs:
...
           docs-folder: docs/
           # The command used to build your documentation.
-          build-command: make html SPHINXOPTS+="-W --keep-going" #  -n"
-          # Run before the build command, you can use this to install system level dependencies, for example with "apt-get update -y && apt-get install -y perl"
-          pre-build-command: |
-            pwd
-            python --version
-            pip install -r requirements.txt
-            pip install -r docs/requirements.txt
+          build-command: make html #"-W --keep-going" #  -n"
```
Comment (skim0119) [inline .github/workflows/ci-sphinx.yml:41]:
```suggestion
          build-command: make html SPHINXOPTS+="-W --keep-going"
```

File: .github/workflows/main.yml [hunk 2]
```diff
@@ -1,110 +1,112 @@
...
 jobs:
   # This workflow contains a single job called "build"
   build:
     # The type of runner that the job will run on
-    runs-on: ${{ matrix.os }} #ubuntu-latest
-
+    runs-on: ${{ matrix.os }}
     strategy:
       matrix:
+        python-version: ["3.7", "3.8", "3.9", "3.10"]
```
Comment (skim0119) [inline .github/workflows/main.yml:23]: why remove 3.6? was there any dependency issue? Also we should have windows check before extending 3.9 and 3.10

**Review by skim0119 — CHANGES_REQUESTED** [review]
LGTM. Some question on supported python version. If we want to include 3.9 and 3.10, we should include the CI for windows together with this PR. Or, we can be vague and just say `python 3.8+`

**Comment by skim0119** [comment]
@bhosale2 Since you are the only one working on this branch, can you rebase and squash those `attempts` commits?

File: .github/workflows/main.yml (hunk 2 above)
Comment (bhosale2) [reply .github/workflows/main.yml:23]: yes, `myst-parser was having issues with 3.6` see below

Screen Shot 2022-07-17 at 11 46 03 PM

I did check windows, and the CI was failing for some test related to pickle

**Comment by skim0119** [comment]
I'll finish the review and merge after I try to add `3.6` (trying to find a way to ignore python-version for extra dependencies).

**Review by armantekinalp — APPROVED** [review]
LGTM

**Comment by skim0119** [comment]
Okay, looks like `mypy` is not the only conflicting dependency. I see `coverage` and `pytest` also shifted their python requirement to be `3.7+`, and I guess there are more.

**Review by skim0119 — APPROVED** [review]
LGTM

**Pull request merged**